File: backend/internal/transport/http/exports.go
package httpapi

import (
	"encoding/csv"
	"net/http"
	"strconv"
	"time"
)

const csvTime = time.RFC3339

func (h Handler) exportApplicationsCSV(w http.ResponseWriter, r *http.Request) {
	ctx := r.Context()

	apps, err := h.applications.List(ctx)
	if err != nil {
		h.writeServiceError(w, err)
		return
	}
	companies, err := h.companies.List(ctx)
	if err != nil {
		h.writeServiceError(w, err)
		return
	}
	companyName := make(map[string]string, len(companies))
	for _, c := range companies {
		companyName[c.ID] = c.Name
	}

	writeCSVHeaders(w, "applications.csv")
	cw := csv.NewWriter(w)
	defer cw.Flush()

	_ = cw.Write([]string{
		"id", "company_id", "company_name", "resume_version_id", "title",
		"role_track", "source", "status", "location", "employment_type",
		"job_url", "applied_at", "deadline_at", "notes", "created_at", "updated_at",
	})

	for _, a := range apps {
		_ = cw.Write([]string{
			a.ID,
			a.CompanyID,
			companyName[a.CompanyID],
			derefString(a.ResumeVersionID),
			a.Title,
			a.RoleTrack,
			derefString(a.Source),
			a.Status,
			derefString(a.Location),
			derefString(a.EmploymentType),
			derefString(a.JobURL),
			formatTimePtr(a.AppliedAt),
			formatTimePtr(a.DeadlineAt),
			derefString(a.Notes),
			a.CreatedAt.UTC().Format(csvTime),
			a.UpdatedAt.UTC().Format(csvTime),
		})
	}
}

func (h Handler) exportContactsCSV(w http.ResponseWriter, r *http.Request) {
	ctx := r.Context()

	contacts, err := h.contacts.List(ctx)
	if err != nil {
		h.writeServiceError(w, err)
		return
	}
	companies, err := h.companies.List(ctx)
	if err != nil {
		h.writeServiceError(w, err)
		return
	}
	companyName := make(map[string]string, len(companies))
	for _, c := range companies {
		companyName[c.ID] = c.Name
	}

	writeCSVHeaders(w, "contacts.csv")
	cw := csv.NewWriter(w)
	defer cw.Flush()

	_ = cw.Write([]string{
		"id", "company_id", "company_name", "name", "role", "email",
		"linkedin_url", "relationship", "notes", "created_at", "updated_at",
	})

	for _, c := range contacts {
		_ = cw.Write([]string{
			c.ID,
			c.CompanyID,
			companyName[c.CompanyID],
			c.Name,
			derefString(c.Role),
			derefString(c.Email),
			derefString(c.LinkedinURL),
			derefString(c.Relationship),
			derefString(c.Notes),
			c.CreatedAt.UTC().Format(csvTime),
			c.UpdatedAt.UTC().Format(csvTime),
		})
	}
}

func (h Handler) exportRemindersCSV(w http.ResponseWriter, r *http.Request) {
	ctx := r.Context()

	reminders, err := h.reminders.List(ctx)
	if err != nil {
		h.writeServiceError(w, err)
		return
	}

	writeCSVHeaders(w, "reminders.csv")
	cw := csv.NewWriter(w)
	defer cw.Flush()

	_ = cw.Write([]string{
		"id", "application_id", "contact_id", "title", "description", "due_at",
		"status", "retry_count", "last_error", "delivered_at", "created_at", "updated_at",
	})

	for _, rm := range reminders {
		_ = cw.Write([]string{
			rm.ID,
			rm.ApplicationID,
			derefString(rm.ContactID),
			rm.Title,
			derefString(rm.Description),
			rm.DueAt.UTC().Format(csvTime),
			rm.Status,
			strconv.FormatInt(int64(rm.RetryCount), 10),
			derefString(rm.LastError),
			formatTimePtr(rm.DeliveredAt),
			rm.CreatedAt.UTC().Format(csvTime),
			rm.UpdatedAt.UTC().Format(csvTime),
		})
	}
}

func writeCSVHeaders(w http.ResponseWriter, filename string) {
	w.Header().Set("Content-Type", "text/csv; charset=utf-8")
	w.Header().Set("Content-Disposition", `attachment; filename="`+filename+`"`)
	w.Header().Set("Cache-Control", "no-store")
}

func derefString(s *string) string {
	if s == nil {
		return ""
	}
	return *s
}

func formatTimePtr(t *time.Time) string {
	if t == nil {
		return ""
	}
	return t.UTC().Format(csvTime)
}

File: backend/src/routes/exports.ts
import type { FastifyPluginCallbackZod } from "fastify-type-provider-zod";
import type { FastifyReply } from "fastify";
import type { ApiServices } from "../services.js";

type ExportServices = Pick<
  ApiServices,
  "applications" | "companies" | "contacts" | "reminders"
>;
export function exportRoutes(
  services: ExportServices,
): FastifyPluginCallbackZod {
  return (app, _options, done) => {
    app.get(
      "/exports/applications.csv",
      { schema: { tags: ["Exports"] } },
      async (_request, reply) => {
        const [items, companies] = await Promise.all([
          services.applications.list(),
          services.companies.list(),
        ]);
        const names = new Map(companies.map((c) => [c.id, c.name]));
        return csv(reply, "applications.csv", [
          [
            "id",
            "company_id",
            "company_name",
            "resume_version_id",
            "title",
            "role_track",
            "source",
            "status",
            "location",
            "employment_type",
            "job_url",
            "applied_at",
            "deadline_at",
            "notes",
            "created_at",
            "updated_at",
          ],
          ...items.map((a) => [
            a.id,
            a.companyId,
            names.get(a.companyId) ?? "",
            a.resumeVersionId ?? "",
            a.title,
            a.roleTrack,
            a.source ?? "",
            a.status,
            a.location ?? "",
            a.employmentType ?? "",
            a.jobUrl ?? "",
            date(a.appliedAt),
            date(a.deadlineAt),
            a.notes ?? "",
            date(a.createdAt),
            date(a.updatedAt),
          ]),
        ]);
      },
    );
    app.get(
      "/exports/contacts.csv",
      { schema: { tags: ["Exports"] } },
      async (_request, reply) => {
        const [items, companies] = await Promise.all([
          services.contacts.list(),
          services.companies.list(),
        ]);
        const names = new Map(companies.map((c) => [c.id, c.name]));
        return csv(reply, "contacts.csv", [
          [
            "id",
            "company_id",
            "company_name",
            "name",
            "role",
            "email",
            "linkedin_url",
            "relationship",
            "notes",
            "created_at",
            "updated_at",
          ],
          ...items.map((c) => [
            c.id,
            c.companyId,
            names.get(c.companyId) ?? "",
            c.name,
            c.role ?? "",
            c.email ?? "",
            c.linkedinUrl ?? "",
            c.relationship ?? "",
            c.notes ?? "",
            date(c.createdAt),
            date(c.updatedAt),
          ]),
        ]);
      },
    );
    app.get(
      "/exports/reminders.csv",
      { schema: { tags: ["Exports"] } },
      async (_request, reply) =>
        csv(reply, "reminders.csv", [
          [
            "id",
            "application_id",
            "contact_id",
            "title",
            "description",
            "due_at",
            "status",
            "retry_count",
            "last_error",
            "delivered_at",
            "created_at",
            "updated_at",
          ],
          ...(await services.reminders.list()).map((r) => [
            r.id,
            r.applicationId,
            r.contactId ?? "",
            r.title,
            r.description ?? "",
            date(r.dueAt),
            r.status,
            String(r.retryCount),
            r.lastError ?? "",
            date(r.deliveredAt),
            date(r.createdAt),
            date(r.updatedAt),
          ]),
        ]),
    );
    done();
  };
}
function csv(reply: FastifyReply, filename: string, rows: string[][]) {
  return reply
    .header("content-type", "text/csv; charset=utf-8")
    .header("content-disposition", `attachment; filename="${filename}"`)
    .header("cache-control", "no-store")
    .send(
      rows
        .map((row) =>
          row
            .map((cell) =>
              /[",\r\n]/.test(cell) ? `"${cell.replaceAll('"', '""')}"` : cell,
            )
            .join(","),
        )
        .join("\n") + "\n",
    );
}
function date(value: Date | null): string {
  return value === null ? "" : value.toISOString().replace(/\.\d{3}Z$/, "Z");
}
